proposals/jsr-305-custom-nullability-qualifiers.md

```diff
@@ -0,0 +1,321 @@
+# JSR-305 custom nullability qualifiers
+
+* **Type**: Design proposal
+* **Author**: Denis Zharkov
+* **Contributors**: Andrey Breslav, Leonid Stashevsky
+* **Status**: Submitted
+* **Prototype**: In progress
+* **Discussion**: [KEEP-79](https://github.com/Kotlin/KEEP/issues/79)
+
+## Summary
+
+- Support loading more precise type information for Java declarations marked 
+with custom nullability annotations based on [JSR-305](https://jcp.org/en/jsr/detail?id=305)
+    - Both `@TypeQualifierNickname` and `@TypeQualifierDefault` should be 
+    supported
+- Introduce `@Migration` annotation and additional compiler flags to allow a 
+library maintainer and its users to control the way how these new annotations
+affect the compiler's behavior. Namely, they may lead to errors, warnings on 
+usages from Kotlin or just be ignored
+
+## Motivation
+
+- It might be too verbose to annotate each signature part in a library API, so
+`TypeQualifierDefault` is the solution. One can simply declare that all types
+in a package are not nullable by default, and additionally annotate nullable 
+parts where it's necessary (or vice versa)
+- The naming of default JSR-305 annotation may be rather confusing:
+`javax.annotation.Nullable` does not actually have the same meaning as nullable
+types in Kotlin, it only means that nullability is unknown unlike the
+`javax.annotation.CheckForNull`. So a library maintainer may want to introduce 
+its own `Nullable` annotation as a type qualifier nickname to `CheckForNull`
+with a more precise/clear meaning. 
+- Annotating already existing parts of a library API is a source incompatible
+change in Kotlin, thus in some cases, it's worth migrating users smoothly:
+    - Release a version of the library where there will be warnings on
+    incorrect usages of the newly annotated parts
+    - Turn these warnings into errors for the next release
+
+## Description
+
+By default all reference types from Java are perceived by the Kotlin compiler
+as [flexible](https://github.com/JetBrains/kotlin/blob/master/spec-docs/flexible-java-types.md), 
+it means that they can be used both as nullable and not-nullable. 
+
+In addition, Java types declarations may be annotated with some of the known 
+annotations to make them have [more precise types](https://github.com/JetBrains/kotlin/blob/master/spec-docs/flexible-java-types.md#more-precise-type-information-from-annotations)
+when used from Kotlin. Let's call this kind of annotating as types enhancement.
+
+Current proposal is supposed to introduce additional ways of types enhancement
+and instruments to control their migration status.
+
+### Type qualifier nicknames
```

Comment:
This part is very hard to understand for people having no background in how JSR-305 is supposed to work. Given that there's no comprehensive doc on JSR-305, I think we should include a brief description of its relevant parts

Reply:
Having read a bit further, I realize that this part is giving the necessary background, but it's not clear upfront. We should add a comment on this, along the lines of "JSR 305 has this and that semantics, and we support the following part of it"